File: programs/js/recursion/mergeSort.js
class Solution {
    merge(arr, l, m, h, output) {
        console.log(output);
        let left = l;
        let right = m + 1;
        const temp = [];

        while (left <= m && right <= h) {
            if (arr[left] <= arr[right]) {
                temp.push(arr[left]);
                left += 1;
            } else {
                temp.push(arr[right]);
                right += 1;
            }
        }

        while (left <= m) {
            temp.push(arr[left]);
            left += 1;
        }

        while (right <= h) {
            temp.push(arr[right]);
            right += 1;
        }
        console.log("temp", temp);
        for (let i = 0; i < temp.length; i++) {
            output[l + i] = temp[i];
        }
    }

    mergeSort(arr, l, h, output) {
        if (l >= h) {
            return;
        }

        const m = Math.floor((l + h) / 2);
        this.mergeSort(arr, l, m, output);
        this.mergeSort(arr, m + 1, h, output);
        this.merge(arr, l, m, h, output);
    }
}

const input = [3, 1, 2, 8, 2, 1, 5, 6, 3, 8, 7];
const output = new Array(input.length);
const solution = new Solution();
solution.mergeSort(input, 0, input.length - 1, output);
// console.log(output);  

File: programs/Python/recursion/merge_sort.py
class Solution:
    def __init__(self):
        pass

    def merge(self, arr, l, m, h, output):
        left = l
        right = m + 1
        temp = []
        while left <= m and right <= h:
            if arr[left] < arr[right]:
                temp.append(arr[left])
                left += 1
            else:
                temp.append(arr[right])
                right += 1
        while left <= m:
            temp.append(arr[left])
            left += 1
        while right <= h:
            temp.append(arr[right])
            right += 1

        # Assign the sorted elements back to the output list
        for i in range(len(temp)):
            output[l + i] = temp[i]

    def merge_sort(self, arr, l, h, output):
        if l < h:
            m = (l + h) // 2
            self.merge_sort(arr, l, m, output)
            self.merge_sort(arr, m + 1, h, output)
            self.merge(arr, l, m, h, output)

if __name__ == "__main__":
    input = [3, 1, 2, 8, 2, 1, 5, 6, 3, 8, 7]
    output = input[:]  # Make a copy of the input list
    solution = Solution()
    solution.merge_sort(input, 0, len(input) - 1, output)
    print(output)
